xain/grpc/participant.py

```diff
@@ -65,34 +118,240 @@ def end_training(
+#     print(f"batch_size: {FLAGS.batch_size}")
+#     print(f"partition_iden: {FLAGS.partition_iden}")
+#     go(init_participant())
+
+
+# if __name__ == "__main__":
+#     flags.mark_flag_as_required("model_name")
+#     flags.mark_flag_as_required("dataset_name")
+#     flags.mark_flag_as_required("batch_size")
+#     flags.mark_flag_as_required("partition_iden")
+#     app.run(main)
+
+
+class StateRecord:
+    """Thread-safe record of Participant state and round number.
+    """
+
+    def __init__(self, state=ParState.WAITING_FOR_SELECTION, round=0):
+        self.cv = threading.Condition()
+        self.round = round
+        self.state = state
+
+    def lookup(self):
+        """Looks up the state and round number.
+
+        Returns:
+            :obj:`Tuple[ParState, int]`: State and round number
+        """
+        with self.cv:
+            return self.state, self.round
+
+    def update(self, state):
+        """Updates state.
+
+        Args:
+            state (:obj:`ParState`): State to update to.
+        """
+        with self.cv:
+            self.state = state
+            self.cv.notify()
+
+    def wait_until_selected_or_done(self):
+        """Waits until Participant is in the state of having been selected for training
+        (or is completely done).
+
+        Returns:
+            :obj:`ParState`: New state Participant is in.
+        """
+        with self.cv:
+            self.cv.wait_for(lambda: self.state in {ParState.TRAINING, ParState.DONE})
+            # which one was it?
+            return self.state
+
+    def wait_until_next_round(self):
+        """Waits until Participant is in a state indicating the start of the next round
+        of training.
+
+        Returns:
+            :obj:`ParState`: New state Participant is in.
+        """
+        with self.cv:
+            self.cv.wait_for(
+                lambda: self.state
+                in {ParState.TRAINING, ParState.WAITING_FOR_SELECTION, ParState.DONE}
+            )
+            # which one was it?
+            return self.state
+
+
+def transit(st, beat_reply):
+    """Participant state transition function on a heartbeat response. Updates the
+    state record `st`.
+
+    Args:
+        st (obj:`StateRecord`): Participant state record to update.
+        beat_reply (obj:`coordinator_pb2.HeartbeatReply`): Heartbeat from Coordinator.
+    """
+    msg, r = beat_reply.state, beat_reply.round
+    with st.cv:
+        if st.state == ParState.WAITING_FOR_SELECTION:
+            if msg == coordinator_pb2.State.ROUND:
+                st.state = ParState.TRAINING
+                st.round = r
+                st.cv.notify()
+            elif msg == coordinator_pb2.State.FINISHED:
+                st.state = ParState.DONE
+                st.cv.notify()
+        elif st.state == ParState.POST_TRAINING:
+            if msg == coordinator_pb2.State.STANDBY:
+                # not selected
+                st.state = ParState.WAITING_FOR_SELECTION
```

Comment:
There is a problem with these state transtions.

Imagine the case where a  participant goes down at the middle of a round and eventually a new one connects. During this time the coordinator with have the following state transitions: `ROUND -> STANDBY -> ROUND`. They all happen in the same round so the round number remains the same.

If this happens while a participant is in the `POST_TRAINING` state the participant with transition to `WAITING_FOR_SELECTION` and then `TRAINING` possibly more than once in the same round.

I think that every transition into `TRAINING` should check whether the participant has already trained in the round or not.

Reply:
Yes. To reiterate what's said, in this situation the coordinator decides to "pause" the round, by telling all participants `STANDBY`. Later it resumes it by sending `ROUND` again (with the same round number as before). So when a participant receives `STANDBY` it shouldn't necessarily assume that the round is already over (as it currently does). Yes, as suggested it could check the round number at this point to make sure it doesn't repeat work already done. But equally, the coordinator (which has already received the update from it) could just continue to tell this participant `STANDBY` - this would be part of the participant-selection logic in the coordinator. I propose postponing an immediate "quick-fix" because
1. I consider round pausing / resuming not critical at this point in time especially as we haven't considered all the fault scenarios very much anyway.
2. This problem should be resolved by a proposal to reduce the participant state machine (which _will_ check round numbers as suggested) and / or participant-selection logic in the coordinator.